File: algorithms/two_sum/two_sum.js
// O(n)
function twoNumberSum(array, targetSum) {
  const index = {};
  for (let i = 0; i < array.length; i++) {
    const possibleMatch = targetSum - array[i];
    const indexedMatch = index[possibleMatch];
    if (indexedMatch == null) {
      index[array[i]] = i;
    } else {
      return [array[indexedMatch], array[i]]; // remove "array[" "]" if you want to get index only
    }
  }
  return [];
}

console.log(twoNumberSum([3, 5, 2, -4, 8, 11], 19));

module.exports = twoNumberSum;

File: algorithms/two_sum/two_sum.py
# O(n) time, O(log(n)) space
def twoNumberSum(arr, targetSum):
    index = {}
    for i in range(len(arr)):
        possibleMatch = targetSum - arr[i]
        indexedMatch = index.get(possibleMatch)
        if indexedMatch is None:
            index[arr[i]] = i
        else:
            return [indexedMatch, i]  # Add arr[ ] to return numbers
    return []


print(twoNumberSum([3, 5, 2, -4, 8, 11], 19))
